src/Microsoft.ML.Data/StaticPipe/TrainerEstimatorReconciler.cs

```diff
@@ -0,0 +1,335 @@
+                if (inputNames.Values.Any(old2New.ContainsKey))
+                {
+                    var newInputNames = new Dictionary<PipelineColumn, string>();
+                    foreach (var p in inputNames)
+                        newInputNames[p.Key] = old2New.ContainsKey(p.Value) ? old2New[p.Value] : p.Value;
+                    inputNames = newInputNames;
+                }
+                result = new CopyColumnsEstimator(env, old2New.Select(p => (p.Key, p.Value)).ToArray());
+            }
+
+            // Map the inputs to the names.
+            string[] mappedInputNames = _inputs.Select(c => inputNames[c]).ToArray();
+            // Finally produce the trainer.
+            var trainerEst = ReconcileCore(env, mappedInputNames);
+            if (result == null)
+                result = trainerEst;
+            else
+                result = result.Append(trainerEst);
+
+            // OK. Now handle the final renamings from the fixed names, to the desired names, in the case
+            // where the output was desired, and a renaming is even necessary.
+            var toRename = new List<(string source, string name)>();
+            foreach ((PipelineColumn outCol, string fixedName) in Outputs.Zip(_outputNames, (c, n) => (c, n)))
+            {
+                if (outputNames.TryGetValue(outCol, out string desiredName))
+                    toRename.Add((fixedName, desiredName));
+                else
+                    env.Assert(!toOutput.Contains(outCol));
+            }
+            // Finally if applicable handle the renaming back from the temp names to the original names.
+            foreach (var p in old2New)
+                toRename.Add((p.Value, p.Key));
+            if (toRename.Count > 0)
+                result = result.Append(new CopyColumnsEstimator(env, toRename.ToArray()));
+
+            return result;
+        }
+
+        /// <summary>
+        /// A reconciler for regression capable of handling the most common cases for regression.
+        /// </summary>
+        public sealed class Regression : TrainerEstimatorReconciler
+        {
+            /// <summary>
+            /// The delegate to parameterize the <see cref="Regression"/> instance.
+            /// </summary>
+            /// <param name="env">The environment with which to create the estimator</param>
+            /// <param name="label">The label column name</param>
+            /// <param name="features">The features column name</param>
+            /// <param name="weights">The weights column name, or <c>null</c> if the reconciler was constructed with <c>null</c> weights</param>
+            /// <returns>Some sort of estimator producing columns with the fixed name <see cref="DefaultColumnNames.Score"/></returns>
+            public delegate IEstimator<ITransformer> EstimatorMaker(IHostEnvironment env, string label, string features, string weights);
```

Comment:
>Maker [](start = 62, length = 5)

Either 'builder' or 'factory' seems a bit more common for a factory method like this

Reply:
Let's do factory I guess.

---
In reply to: [215366624](https://github.com/dotnet/machinelearning/pull/837#discussion_r215366624) [](ancestors = 215366624)